src/fklearn/causal/cate_learning/meta_learners.py

```diff
@@ -0,0 +1,233 @@
+import copy
+import inspect
+from typing import List, Tuple
+
+import numpy as np
+import pandas as pd
+from fklearn.common_docstrings import (learner_pred_fn_docstring,
+                                       learner_return_docstring)
+from fklearn.training.classification import lgbm_classification_learner
+from fklearn.training.pipeline import build_pipeline
+from fklearn.types import LearnerFnType, LearnerReturnType, PredictFnType
+from toolz import curry
+
+TREATMENT_FEATURE = "is_treatment"
+
+
+def _append_treatment_feature(features: list, treatment_feature: str) -> list:
+    return features + [treatment_feature]
+
+
+def _get_learner_features(learner: LearnerFnType) -> list:
+    return inspect.signature(learner).parameters["features"].default
+
+
+def _get_unique_treatments(
+    df: pd.DataFrame, treatment_col: str, control_name: str
+) -> list:
+    return [col for col in df[treatment_col].unique() if col != control_name]
+
+
+def _filter_by_treatment(
+    df: pd.DataFrame, treatment_col: str, treatment_name: str, control_name: str
+) -> pd.DataFrame:
+    treatment_control_df = df.loc[
+        (df[treatment_col] == treatment_name) | (df[treatment_col] == control_name), :
+    ]
+    return treatment_control_df
+
+
+def _create_treatment_flag(
+    df: pd.DataFrame, treatment_col: str, treatment_name: str
+) -> pd.DataFrame:
+    df = df.copy()
+    treatment_flag = np.where(df[treatment_col] == treatment_name, 1.0, 0.0)
+    df[TREATMENT_FEATURE] = treatment_flag
+
+    return df
+
+
+def _fit_by_treatment(
+    df: pd.DataFrame,
+    learner: LearnerFnType,
+    treatment_col: str,
+    control_name: str,
+    treatments: list,
+) -> Tuple[dict, dict]:
+    fitted_learners = {}
+    learners_logs = {}
+    for treatment in treatments:
+        treatment_control_df = _filter_by_treatment(
+            df=df,
+            treatment_col=treatment_col,
+            treatment_name=treatment,
+            control_name=control_name,
+        )
+        treatment_control_df = _create_treatment_flag(
+            df=treatment_control_df,
+            treatment_col=treatment_col,
+            treatment_name=treatment,
+        )
+        learner_fcn, _, learner_log = learner(treatment_control_df)
+        fitted_learners[treatment] = learner_fcn
+        learners_logs[treatment] = learner_log
+
+    return fitted_learners, learners_logs
+
+
+def _predict_for_control(
+    df: pd.DataFrame, learner_fcn: PredictFnType, prediction_column: str
+) -> np.array:
+    control_flag = np.zeros(df.shape[0])
+    df[TREATMENT_FEATURE] = control_flag
+    control_pred_df = learner_fcn(df)
+    return control_pred_df[prediction_column].values
+
+
+def _predict_for_treatment(
+    df: pd.DataFrame, learner_fcn: PredictFnType, prediction_column: str
+) -> np.array:
+    treatment_flag = np.ones(df.shape[0])
+    df[TREATMENT_FEATURE] = treatment_flag
+    treatment_pred_df = learner_fcn(df)
+    return treatment_pred_df[prediction_column].values
```

Comment:
These are very similar. I think it would be better to have a _predict_for_treatment_value that would receive the value (0 or 1) as an argument.

Reply:
✅